compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmReflectSymbols.kt

```diff
@@ -0,0 +1,149 @@
+import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
+import org.jetbrains.kotlin.ir.types.defaultType
+import org.jetbrains.kotlin.ir.types.makeNullable
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+class JvmReflectSymbols(val context: JvmBackendContext) {
+    private val irBuiltIns: IrBuiltIns = context.irBuiltIns
+
+    private val javaLangReflect: FqName = FqName("java.lang.reflect")
+
+    private val javaLangReflectPackage: IrPackageFragment =
+        IrExternalPackageFragmentImpl.createEmptyExternalPackageFragment(context.state.module, javaLangReflect)
+
+    val javaLangReflectField: IrClassSymbol =
+        createJavaLangReflectClass(FqName("java.lang.reflect.Field")) { klass ->
+            klass.addFunction("setAccessible", irBuiltIns.unitType).apply {
+                addValueParameter("isAccessible", irBuiltIns.booleanType)
+            }
+            klass.addFunction("get", irBuiltIns.anyNType).apply {
+                addValueParameter("receiver", irBuiltIns.anyNType)
+            }
+            klass.addFunction("set", irBuiltIns.unitType).apply {
+                addValueParameter("receiver", irBuiltIns.anyNType)
+                addValueParameter("value", irBuiltIns.anyNType)
+            }
+        }
+
+    val javaLangReflectMethod: IrClassSymbol =
+        createJavaLangReflectClass(FqName("java.lang.reflect.Method")) { klass ->
+            klass.addFunction("setAccessible", irBuiltIns.unitType).apply {
+                addValueParameter("isAccessible", irBuiltIns.booleanType)
+            }
+            klass.addFunction("invoke", irBuiltIns.anyNType).apply {
+                addValueParameter("receiver", irBuiltIns.anyNType)
+                addValueParameter {
+                    name = Name.identifier("args")
+                    type = irBuiltIns.arrayClass.typeWith(irBuiltIns.anyNType)
+                    varargElementType = irBuiltIns.anyNType
+                }
+            }
+        }
+
+    val javaLangReflectConstructor: IrClassSymbol =
+        createJavaLangReflectClass(FqName("java.lang.reflect.Constructor")) { klass ->
+            klass.addFunction("setAccessible", irBuiltIns.unitType).apply {
+                addValueParameter("isAccessible", irBuiltIns.booleanType)
+            }
+            klass.addFunction("newInstance", irBuiltIns.anyNType).apply {
+                addValueParameter {
+                    name = Name.identifier("args")
+                    type = irBuiltIns.arrayClass.typeWith(irBuiltIns.anyNType)
+                    varargElementType = irBuiltIns.anyNType
+                }
+            }
+        }
+
+    init {
+        val klass = context.ir.symbols.javaLangClass.owner
+        if (!klass.functions.any { it.name.asString() == "getDeclaredMethod" }) {
```

Comment:
@udalov Is this "too defensive"? What I am thinking is by doing this mutation of JvmSymbols, I guess it's possible to create an inconsistent state if `JvmReflectSymbols` is instantiated a number of times.

Reply:
Nevermind, figured lazy initialization was a cleaner, clearer way of guarding for the same thing.